apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/SizedBufferingConsumer.cs

```diff
@@ -61,43 +63,115 @@ private SizedBufferingConsumer(IConsumer<T> consumer, ISizer<T> sizer, int buffe
+                }
+            }
+            finally
+            {
+                _semaphore.Release();
+            }
         }
 
         /// <inheritdoc />
-        public void Receive(IEnumerable<T> items)
+        public async Task ReceiveAsync(
+            IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            lock (_mutex)
+            try
             {
+                await _semaphore.WaitAsync(cancellationToken);
                 foreach (T item in items)
                 {
                     _size += _sizer.GetSize(item);
                     _items.Add(item);
                     if (_size >= _bufferSize)
                     {
-                        Flush();
+                        await FlushesAsyncWithSemaphoreHeld(cancellationToken);
                     }
                 }
             }
+            finally
+            {
+                _semaphore.Release();
+            }
         }
 
         /// <inheritdoc />
-        public void Flush()
+        public void Flushes()
         {
-            IList<T> old;
-            lock (_mutex)
+            try
+            {
+                _semaphore.Wait();
+                FlushesWithSemaphoreHeld();
+            }
+            finally
+            {
+                _semaphore.Release();
+            }
+        }
+
+        /// <inheritdoc />
+        public async Task FlushesAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            try
             {
-                old = _items;
-                _items = new List<T>();
-                _size = 0;
+                await _semaphore.WaitAsync();
+                await FlushesAsyncWithSemaphoreHeld(cancellationToken);
             }
+            finally
+            {
+                _semaphore.Release();
+            }
+        }
+
+        /// <summary>
+        /// Flushes the buffer. The caller is responsible for ensuring the semaphore is held.
+        /// </summary>
+        private void FlushesWithSemaphoreHeld()
+        {
+            IList<T> old;
+            old = _items;
+            _items = new List<T>();
+            _size = 0;
 
             if (old.Count != 0)
             {
                 _consumer.Receive(old);
             }
         }
+
+        /// <summary>
+        /// Flushes the buffer asynchronously. The caller is responsible for ensuring the semaphore is held.
+        /// </summary>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        private async Task FlushesAsyncWithSemaphoreHeld(CancellationToken cancellationToken)
+        {
+            IList<T> old;
+            old = _items;
+            _items = new List<T>();
+            _size = 0;
+
+            if (old.Count != 0)
```

Comment:
Given that this has to be called within a semaphore, isn't it simpler just to have

```csharp
if (old.Count == 0)
{
    return;
}
```

at the start of the method? We don't need to do the list-swapping in that case, right?